File: frontend/src/services/volunteer.ts
import { apiClient } from "./apiClient";

export interface Volunteer {
  id: string;
  user_id: string;
  name: string;
  role: string;
  status: string;
  zone: string | null;
  languages: string;
  phone: string | null;
  created_at: string;
}

export interface VolunteerTask {
  id: string;
  volunteer_id: string | null;
  task_type: string;
  description: string;
  zone: string | null;
  priority: string;
  status: string;
  assigned_at: string | null;
  completed_at: string | null;
  created_at: string;
}

export interface VolunteerDashboard {
  total: number;
  on_shift: number;
  available: number;
  active_tasks: number;
  volunteers: Volunteer[];
  tasks: VolunteerTask[];
}

export const fetchVolunteerDashboard = () =>
  apiClient<VolunteerDashboard>("/volunteer/dashboard");

export const fetchVolunteers = (role?: string) =>
  apiClient<Volunteer[]>(`/volunteer/volunteers${role ? `?role=${role}` : ""}`);

export const createVolunteer = (data: { name: string; role: string; zone?: string; languages?: string; phone?: string }) =>
  apiClient<Volunteer>("/volunteer/volunteers", { method: "POST", body: JSON.stringify(data) });

export const updateVolunteerStatus = (id: string, data: { status: string; zone?: string }) =>
  apiClient<Volunteer>(`/volunteer/volunteers/${id}`, { method: "PATCH", body: JSON.stringify(data) });

export const fetchTasks = (status?: string) =>
  apiClient<VolunteerTask[]>(`/volunteer/tasks${status ? `?status=${status}` : ""}`);

export const updateTask = (id: string, data: { status: string; volunteer_id?: string }) =>
  apiClient<VolunteerTask>(`/volunteer/tasks/${id}`, { method: "PATCH", body: JSON.stringify(data) });

File: backend/app/models/volunteer.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import Column, String, DateTime, Text
from app.models.base import Base
import enum


class VolunteerRole(str, enum.Enum):
    gate_ops = "gate_ops"
    concierge = "concierge"
    transit_support = "transit_support"
    accessibility = "accessibility"
    emergency_response = "emergency_response"


class VolunteerStatus(str, enum.Enum):
    available = "available"
    on_shift = "on_shift"
    on_break = "on_break"
    off_duty = "off_duty"


class VolunteerTaskStatus(str, enum.Enum):
    assigned = "assigned"
    in_progress = "in_progress"
    completed = "completed"
    cancelled = "cancelled"


class Volunteer(Base):
    __tablename__ = "volunteers"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id = Column(String, nullable=False, index=True)
    name = Column(String(200), nullable=False)
    role = Column(String(50), nullable=False, default=VolunteerRole.concierge.value)
    status = Column(String(20), nullable=False, default=VolunteerStatus.available.value)
    zone = Column(String(10), nullable=True)
    languages = Column(String(500), default="en")
    phone = Column(String(20), nullable=True)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))


class VolunteerTask(Base):
    __tablename__ = "volunteer_tasks"

    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    volunteer_id = Column(String, nullable=True, index=True)
    task_type = Column(String(50), nullable=False)
    description = Column(Text, nullable=False)
    zone = Column(String(10), nullable=True)
    priority = Column(String(10), default="medium")
    status = Column(String(20), default=VolunteerTaskStatus.assigned.value)
    assigned_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    completed_at = Column(DateTime, nullable=True)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
